llvm/lib/Target/RISCV/RISCVInsertVSETVLI.cpp

```diff
@@ -964,6 +965,66 @@ void RISCVInsertVSETVLI::insertVSETVLI(MachineBasicBlock &MBB,
       .addImm(TypeI);
 }
 
+void RISCVInsertVSETVLI::insertVSETVLIForCOPY(MachineBasicBlock &MBB) {
+  auto findRegSequence = [] (const MachineBasicBlock& BB,
+                             MachineBasicBlock::iterator I,
+                             Register DstReg) {
+    for (; I != BB.end(); ++I)
+      if (I->getOpcode() == TargetOpcode::REG_SEQUENCE &&
+          I->getOperand(1).getReg() == DstReg)
+        break;
+    return I;
+  };
+
+  for (auto MII = MBB.begin(); MII != MBB.end(); ++MII) {
+    MachineInstr &MI = *MII;
+    // Check if the COPY node will be converted to PseudoTH_VMV<n>R
+    // later. Set VTYPE and VL for the pseudo node if so.
+    if (!(MI.isCopy() && TII->needVSETVLIForCOPY(MBB, MI)))
+      continue;
+
+    auto DL = MI.getDebugLoc();
+
+    // Always set SEW to 8. Actually no matter what the SEW is the VMV
+    // instruction will behave correctly because it does not do sign
+    // extension and we always set VL to VLMAX.
+    // (It's an ugly solution, though. Maybe we can simulate GCC later.)
+    unsigned SEW = 8;
+
+    auto SavedVL = MRI->createVirtualRegister(&RISCV::GPRRegClass);
+    auto SavedVType = MRI->createVirtualRegister(&RISCV::GPRRegClass);
+    BuildMI(MBB, MI, DL, TII->get(RISCV::CSRRS), SavedVL)
+        .addImm(RISCVSysReg::lookupSysRegByName("VL")->Encoding)
+        .addReg(RISCV::X0);
+    BuildMI(MBB, MI, DL, TII->get(RISCV::CSRRS), SavedVType)
+        .addImm(RISCVSysReg::lookupSysRegByName("VTYPE")->Encoding)
+        .addReg(RISCV::X0);
+
+    auto NewVType = RISCVVType::encodeXTHeadVTYPE(SEW, 1, 1);
+    BuildMI(MBB, MI, DL, TII->get(RISCV::PseudoTH_VSETVLIX0))
+        .addReg(RISCV::X0, RegState::Define | RegState::Dead)
+        .addReg(RISCV::X0)
+        .addImm(NewVType);
```

Comment:
Maybe we can extract the code generating backup-action-restore to a separate function, and reuse them in `emitXWholeLoadStore`.